### ISSUE TITLE
Error adding a new item to repeatable component

### ISSUE THREAD

**Issue by chadwithuhc** [strapi/strapi#11980]
### Describe the bug

I am trying to add a new item to a repeatable component field through an API `PUT /api/[api-name]/:id`. I expected to be able to push a new item on the array without an ID and it will be created, but I receive an error instead. I am not sure if this is a bug or my misunderstanding of the API.

### Steps to reproduce the behavior

Example of Lists Content Type
<img width="919" alt="Example of Lists Content Type" src="https://user-images.githubusercontent.com/259196/147013877-189f5b91-6287-40a1-ad1e-1456be035ac1.png">

1. Inside my `beforeUpdate()` lifecycle hook for my `Lists` API,
2. I try to add a new item to my existing `items` array by using:
```
event.params.data.items.push({
  title: "Washcloths",
  checked: true
})
```
3. I can see it is properly added to the data
```
// event.params.data
{
  title: 'Craigslist',
  items: [ { title: 'Washcloths', checked: true } ],
  updatedAt: 2021-12-22T00:21:36.488Z
}
```
4. Then I get the error `error: Invalid id, expected a string or integer, got [object Object]` once it is run
```
Error: Invalid id, expected a string or integer, got [object Object]
    at node_modules/@strapi/database/lib/entity-manager.js:25:15
```

### Expected behavior

I would expect it to create a new item for me.

### System

- Node.js version: v16.13.0
- NPM version: 8.1.0
- Strapi version: 4.0.0
- Database: sqlite3
- Operating system: Mac OS Monterrey

**Comment by iicdii** [comment]
I did `console.log` in source code and I found the difference between manual data and internal data.
```js
[
  {
    id: 3,
    __pivot: { order: 1, field: 'items', component_type: 'default.items' }
  }, // from internal entity-manager
  { title: 'Washcloths', checked: true } // from beforeUpdate lifecycle
]
```
I think the data from `beforeUpdate` lifecycle should also be converted too in internal code.

Or you can use different approach like below - overriding your core service
```js
// src/api/list/services/list.js

module.exports = createCoreService('api::list.list', ({ strapi }) =>  ({
  async update(entityId, params) {
    params.data.items = (params.data.items || []).concat({
      title: "Washcloths",
      checked: true
    });
    const result = await super.update(entityId, params);

    return result;
  }
}));
```

**Comment by chadwithuhc** [comment]
That solves my problem for now, thanks @iicdii !

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/lifecycle-and-components-inside-collections-type/14003/2

**Comment by osmanorhan** [comment]
This issue still persists.

**Comment by DeploymentService** [comment]
is there a way to convert the data from beforeUpdate format into an internal code and give it an id ? @iicdii

**Comment by iicdii** [comment]
@DeploymentService I don' think that's possible. and also dangerous - you don't know when internal code'd be changed in the future. Currently it's the best way to override the core service like example in comments.

**Comment by WardenCommander** [comment]
I think lifecycles on components would help alot of people and open alot of new possibilities

**Comment by derrickmehaffy** [comment]
This is being discussed internally, is everyone here trying to use components inside of lifecycles?

We are looking at this issue more broadly than just as a bug.

**Comment by Saad-ISAA** [comment]
I was trying to encrypt some fields in the beforeCreate hook. and noticed that components are not part of the data object in the hook

**Comment by derrickmehaffy** [comment]
> I was trying to encrypt some fields in the beforeCreate hook. and noticed that components are not part of the data object in the hook

Yup that's true and I don't know if they will be, but we are talking about possibly introducing lifecycles for components.

Technically at the layer these lifecycles run (db) components are no different than normal content-types. They only become components in the entityService layer and higher and we don't have lifecycles for the entityService layer.

@Convly had the idea to introduce lifecycles at every layer as an option but that will require some discussion/rfcs/time/ect.

I'd like to get some use-cases for lifecycles on components and what all of you are trying to do so I can approach the product managers about this. If your willing to share use-cases please do.

**Comment by Saad-ISAA** [comment]
though i like the idea of having lifecycles for components and since these are almost identical to content-types then even the REST api for components is also somewhat possible.
However components should be in the lifecycles of content-types.
I have many use cases, I'll tell a few:
1) when i have address component in a content type, i need to fill city and country at lifecycle beforeCreate beforeUpdate, coz that is something not being posted from the front end (strict old-school-front-end code) which cannot be modified so i handled that at backend in v3.
2) i have to store some sensitive data in strapi so have to encrypt it beforeCreate. that part was also being done in lifecycles, and since i can't in v4, so handled that in controller side now.

**Comment by derrickmehaffy** [comment]
> REST api for components is also somewhat possible.

At this time we have no plans to introduce this

> However components should be in the lifecycles of content-types.

I won't say it's not possible, but the technical challenges to do so would make that extremely difficult given the current state of the code-base. I find it unlikely the engineering and product teams would be okay with this (I could be wrong) but if we are going to do anything, most likely we could implement the `state` to pass that down to component lifecycles and then back up 

So something like:

- beforeCreate (content-type) with `state`
- beforeCreate (component #1) with shared upstream `state`
- beforeCreate (component #2) with shared upstream `state`
- afterCreate (component #1) with shared upstream `state`
- afterCreate (component #2) with shared upstream `state`
- afterCreate (content-type) with shared `state` from previous steps

**Comment by hashhex** [comment]
maybe so

```
 ../src/index.js

module.exports = {
  bootstrap({ strapi }) {
    strapi.db.lifecycles.subscribe({
      models: ['api::list.list'],
      async afterCreate (list) {
         let { result } = list
         await strapi.entityService.update('api::list.list', result.id, {
            data: {
               title: "Washcloths",
               checked: true
            }
         })
      }
    })
  }
}
```

**Comment by rcdk** [comment]
I would also love to see this addition ;-) Thanks

**Comment by finfa** [comment]
My use case is to conditionally validate internal component fields depending on which model they are used in. I also need to trigger particular logics depending on which component fields have been changed. So I need the values of the component fields within the preUpdate of the models. This worked perfectly in v3. The idea of lifecycle callbacks of individual components would be of no use to me.

**Comment by 88media** [comment]
I use an address component that needs to call some geocoding service in the lifecycle hook.

**Comment by qraxiss** [comment]
You have to create in database repeatable components. And update your api collection data.

```typescript
    let option1 = await strapi.db.query('product.option').create({
        data: {
            value: "XXL"
        }
    })

    let option2 = await strapi.db.query('product.option').create({
        data: {
            value: "XL"
        }
    })

    let variant1 = await strapi.db.query('product.variants').create({
        data: {
            options: [option1, option2],
            name: "size"
        }
    })

    let product = await strapi.db.query('api::product.product').update({
        where :{
            slug: "qraxiss-x"
        },
        populate: {
            variants: "*"
        },
        data : {
            variants: [variant1]
        }
    })
```
Here is the result! have a nice day!
<img width="415" alt="Screenshot 2024-02-25 at 12 38 25" src="https://github.com/strapi/strapi/assets/92579832/91c71033-00ec-4d2f-8dbb-2a0907787053">
